editor/src/doc.rs

```diff
@@ -95,56 +118,179 @@ impl<'l> Doc<'l> {
+    fn execute_ed(&mut self, _cmd: EditorCmd) -> bool {
+        unimplemented!();
+    }
+
+    fn execute_doc(&mut self, cmd: DocCmd) -> bool {
         let undos = match cmd {
-            NavCmd::Left => {
+            DocCmd::Undo => {
+                self.undo(); // TODO: warn if false
+                vec![]
+            }
+            DocCmd::Redo => {
+                self.redo(); // TODO: warn if false
+                vec![]
+            }
+        };
+        self.recent.append(UndoGroup {
+            contains_edit: true,
+            commands: undos,
+        });
+        true
+    }
+
+    fn execute_tree(&mut self, cmd: TreeCmd<'l>) -> bool {
+        if !self.mode.is_tree() {
+            // return false;
+            panic!("tried to execute tree command in text mode")
+        }
+        let undos = match cmd {
+            TreeCmd::Replace(new_ast) => {
+                let i = self.ast.index();
+                self.ast.goto_parent();
+                let old_ast = match self.ast.inner() {
+                    AstKind::Fixed(mut fixed) => {
+                        let old_ast = fixed.replace_child(i, new_ast);
+                        fixed.goto_child(i);
+                        old_ast
+                    }
+                    AstKind::Flexible(mut flexible) => {
+                        let old_ast = flexible.replace_child(i, new_ast);
+                        flexible.goto_child(i);
+                        old_ast
+                    }
+                    _ => panic!("how can a parent not be fixed or flexible?"),
+                };
+                vec![TreeCmd::Replace(old_ast).into()]
+            }
+            TreeCmd::Remove => {
+                let i = self.ast.index();
+                self.ast.goto_parent();
+                match self.ast.inner() {
+                    AstKind::Fixed(mut fixed) => {
+                        // Oops, go back, we can't delete something from a fixed node.
+                        fixed.goto_child(i);
+                        return false;
+                    }
+                    AstKind::Flexible(mut flexible) => {
+                        let old_ast = flexible.remove_child(i);
+                        let num_children = flexible.num_children();
+                        if num_children == 0 {
+                            // Stay at the childless parent
+                            vec![TreeCmd::InsertPrepend(old_ast).into()]
+                        } else if num_children == i {
+                            // We removed the child at the end, so the old index doesn't exist.
+                            flexible.goto_child(i - 1);
+                            vec![TreeCmd::InsertAfter(old_ast).into()]
+                        } else {
+                            // Go back to the same index.
+                            flexible.goto_child(i);
+                            vec![TreeCmd::InsertBefore(old_ast).into()]
+                        }
+                    }
```

Comment:
TODO: should act like backspace, rather than delete.

    num_children==0 -> same
    i==0 -> goto(0); undo(InsertBefore)
    i==1 -> goto(0); undo(InsertAfter)